onnxruntime/core/session/provider_bridge_ort.cc

```diff
@@ -0,0 +1,563 @@
+    return onnxruntime::make_unique<Provider_AttributeProto_Impl>(v_);
+  }
+
+  ::onnx::AttributeProto_AttributeType type() const override { return v_.type(); }
+
+  int ints_size() const override {
+    return v_.ints_size();
+  }
+
+  int64_t ints(int i) const override { return v_.ints(i); }
+  int64_t i() const override { return v_.i(); }
+  float f() const override { return v_.f(); }
+  void set_s(const ::std::string& value) override { v_.set_s(value); }
+  const ::std::string& s() const override { return v_.s(); }
+  void set_name(const ::std::string& value) override { v_.set_name(value); }
+  void set_type(::onnx::AttributeProto_AttributeType value) override { v_.set_type(value); }
+  ::onnx::Provider_TensorProto* add_tensors() override {
+    // Kind of a hack, but the pointer is only valid until the next add_tensors call
+    tensors_ = onnxruntime::make_unique<Provider_TensorProto_Impl>(v_.add_tensors());
+    return tensors_.get();
+  }
+
+  ONNX_NAMESPACE::AttributeProto v_;
+  std::unique_ptr<Provider_TensorProto_Impl> tensors_;
+};
+
+struct Provider_KernelDef_Impl : Provider_KernelDef {
+  Provider_KernelDef_Impl(std::unique_ptr<KernelDef> p) : p_(std::move(p)) {}
+  std::unique_ptr<KernelDef> p_;
+};
+
+struct Provider_KernelDefBuilder_Impl : Provider_KernelDefBuilder {
+  Provider_KernelDefBuilder& SetName(const char* op_name) override {
+    v_.SetName(op_name);
+    return *this;
+  }
+  Provider_KernelDefBuilder& SetDomain(const char* domain) override {
+    v_.SetDomain(domain);
+    return *this;
+  }
+
+  Provider_KernelDefBuilder& SinceVersion(int since_version) override {
+    v_.SinceVersion(since_version);
+    return *this;
+  }
+  Provider_KernelDefBuilder& Provider(const char* provider_type) override {
+    v_.Provider(provider_type);
+    return *this;
+  }
+  Provider_KernelDefBuilder& TypeConstraint(const char* arg_name, MLDataType supported_type) override {
+    v_.TypeConstraint(arg_name, supported_type);
+    return *this;
+  }
+
+  std::unique_ptr<Provider_KernelDef> Build() override {
+    return onnxruntime::make_unique<Provider_KernelDef_Impl>(v_.Build());
+  }
+
+  KernelDefBuilder v_;
+};
+
+struct Provider_NodeArg_Impl : Provider_NodeArg {
+  Provider_NodeArg_Impl(const NodeArg* p) : p_{p} {
+    if (p_->Shape())
+      tensor_shape_proto_.dim_size_ = p_->Shape()->dim_size();
+  }
+
+  const std::string& Name() const noexcept override { return p_->Name(); }
+  const ONNX_NAMESPACE::Provider_TensorShapeProto* Shape() const override { return &tensor_shape_proto_; }
+  virtual ONNX_NAMESPACE::DataType Type() const noexcept override { return p_->Type(); }
+
+  const NodeArg* p_;
+  ONNX_NAMESPACE::Provider_TensorShapeProto tensor_shape_proto_;
+};
+
+struct Provider_Node_Impl : Provider_Node {
+  Provider_Node_Impl(const Node* p) : p_{p} {}
+  ~Provider_Node_Impl() override {
+    for (auto p : input_defs_)
+      delete p;
+    for (auto p : output_defs_)
+      delete p;
+  }
```

Comment:
The Graph instance owns and cleans up all the NodeArg instances in the existing code. Why do we need different ownership semantics here?

Reply:
It's mainly because of the InputDefs() and OutputDefs() functions working with vectors of pointers. There's no way to simpliy convert a vector of unique_ptr to just pointers. Since the wrapper had to allocate wrappers for each of them, I did it manually. It's ugly, and I hope to design it away in the future.